internal/testserver/server.go

```diff
@@ -0,0 +1,135 @@
+package testserver
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/databricks/cli/internal/testutil"
+	"github.com/databricks/cli/libs/testdiff"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type Server struct {
+	*httptest.Server
+	Mux *http.ServeMux
+
+	t   testutil.TestingT
+	ctx context.Context
+
+	// API calls that we expect to be made.
+	calledPatterns map[string]bool
+}
+
+type ApiSpec struct {
+	Method  string `json:"method"`
+	Path    string `json:"path"`
+	Request struct {
+		Headers map[string]string `json:"headers"`
+		Body    json.RawMessage   `json:"body"`
+	} `json:"request"`
+	Response struct {
+		Body json.RawMessage `json:"body"`
+	} `json:"response"`
+}
+
+func New(ctx context.Context, t testutil.TestingT) *Server {
+	mux := http.NewServeMux()
+	server := httptest.NewServer(mux)
+
+	return &Server{
+		Server:         server,
+		Mux:            mux,
+		t:              t,
+		ctx:            ctx,
+		calledPatterns: make(map[string]bool),
+	}
+}
+
+func NewFromConfig(t testutil.TestingT, path string) *Server {
+	content := testutil.ReadFile(t, path)
+	var apiSpecs []ApiSpec
+	err := json.Unmarshal([]byte(content), &apiSpecs)
+	require.NoError(t, err)
+
+	ctx, replacements := testdiff.WithReplacementsMap(context.Background())
+	testdiff.PrepareReplacementOS(t, replacements)
+	testdiff.PrepareReplacementsUUID(t, replacements)
+	testdiff.PrepareReplacementsSemver(t, replacements)
+
+	server := New(ctx, t)
+	for _, apiSpec := range apiSpecs {
+		server.MustHandle(apiSpec)
+	}
+
+	return server
+}
+
+type HandlerFunc func(req *http.Request) (resp any, err error)
+
+func (s *Server) MustHandle(apiSpec ApiSpec) {
+	require.NotEmpty(s.t, apiSpec.Method)
+	require.NotEmpty(s.t, apiSpec.Path)
+
+	pattern := apiSpec.Method + " " + apiSpec.Path
+	s.calledPatterns[pattern] = false
+
+	s.Handle(pattern, func(req *http.Request) (any, error) {
+		for k, v := range apiSpec.Request.Headers {
+			testdiff.AssertEqualStrings(s.t, s.ctx, v, req.Header.Get(k))
+		}
+
+		b, err := io.ReadAll(req.Body)
+		require.NoError(s.t, err)
```

Comment:
It's better not to use require in handlers. For an explanation, see https://github.com/Antonboom/testifylint?tab=readme-ov-file search for "HTTP handlers"

Reply:
Ah, that's really subtle. Thanks!